lzy/channel-manager/src/main/java/ai/lzy/channelmanager/v2/services/SlotsService.java

```diff
@@ -0,0 +1,496 @@
+                        var state = consumer.description().hasStoragePeer() ? State.ACTIVE : State.PENDING;
+
+                        transferDao.create(transferId, peerId, consumer.id(), channelId, state, tx);
+                        transfers.add(new PeerAndTransfer(consumer, transferId));
+                    }
+
+                    tx.commit();
+                    return new ProducerAndConsumerTransfers(prod, transfers);
+                }
+            });
+        } catch (Exception e) {
+            LOG.error("{} Cannot save peer description in db: ", logPrefix, e);
+            throw Status.INTERNAL
+                .withDescription("Cannot save peer description in db")
+                .asRuntimeException();
+        }
+        return result;
+    }
+
+    /**
+     * Atomic consumer creation
+     * If producer not found, mark consumer as not connected
+     * Else mark it connected, create new active transfer and return producer with this transfer
+     * All done in one transaction
+     * @return producer with transfer id or null if producer not found
+     */
+    @Nullable
+    private PeerAndTransfer createConsumer(String channelId, String peerId, PeerDescription peerDesc, String logPrefix,
+                                           Peer.Role role)
+    {
+        final PeerAndTransfer result;
+        try {
+            result = DbHelper.withRetries(LOG, () -> {
+                try (var tx = TransactionHandle.create(storage)) {
+                    var producer = peerDao.findProducer(channelId, tx);
+                    peerDao.create(channelId, peerDesc, role, PeerDao.Priority.PRIMARY, producer != null, tx);
+
+                    String transferId = null;
+
+                    if (producer != null) {
+                        transferId = idGenerator.generate("transfer-");
+
+                        // We are assuming call cannot fail after this transaction
+                        // So we can create transfer already in active state
+                        transferDao.create(
+                            transferId, producer.id(), peerId, channelId, State.ACTIVE, tx);
+                    }
+
+                    tx.commit();
+                    return producer == null ? null : new PeerAndTransfer(producer, transferId);
+                }
+            });
+        } catch (Exception e) {
+            LOG.error("{} Cannot save peer description in db: ", logPrefix, e);
+            throw Status.INTERNAL
+                .withDescription("Cannot save peer description in db")
+                .asRuntimeException();
+        }
+        return result;
+    }
+
+    private record PeerAndTransfer(Peer peer, String transferId) {}
+
+    private record ProducerAndConsumerTransfers(Peer producer, List<PeerAndTransfer> transfers) {}
+
+    @Override
+    public void unbind(LCMS.UnbindRequest request, StreamObserver<LCMS.UnbindResponse> responseObserver) {
```

Comment:
it's better to order methods by visibility: i.e. `public` methods should be placed before the `private` ones